File: formula-java/src/main/java/org/formula/parse/tree/CharacterTokenMatcher.java
package org.formula.parse.tree;

public class CharacterTokenMatcher implements TokenMatcher {
    private final char character;

    public static CharacterTokenMatcher of(char character) {
        return new CharacterTokenMatcher(character);
    }

    @Override
    public boolean matches(char c) {
        return c == character;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        CharacterTokenMatcher that = (CharacterTokenMatcher) o;
        return character == that.character;
    }

    @Override
    public int hashCode() {
        return character;
    }

    public CharacterTokenMatcher(char character) {
        this.character = character;
    }
}

File: formula-cs/Formula/TokenTree/CharacterTokenMatcher.cs
namespace Formula.TokenTree;

internal class CharacterTokenMatcher : ITokenMatcher
{
    private readonly char _character;
    
    public static CharacterTokenMatcher Of(char character) {
        return new CharacterTokenMatcher(character);
    }

    public bool Matches(char c) {
        return c == _character;
    }
    
    public override bool Equals(object? o)
    {
        return o is CharacterTokenMatcher other && Equals(other);
    }

    private bool Equals(CharacterTokenMatcher other)
    {
        return _character == other._character;
    }

    public override int GetHashCode()
    {
        return _character.GetHashCode();
    }

    private CharacterTokenMatcher(char character) {
        _character = character;
    }
}